src/common/page_table.h

```diff
@@ -48,6 +45,43 @@ struct SpecialRegion {
  * mimics the way a real CPU page table works.
  */
 struct PageTable {
+    static constexpr int ATTRIBUTE_BITS = 2;
+
+    class PageInfo {
+    public:
+        [[nodiscard]] u8* Pointer() const noexcept {
+            return ExtractPointer(raw);
+        }
+
+        [[nodiscard]] PageType Type() const noexcept {
+            return ExtractType(raw);
+        }
+
+        [[nodiscard]] std::pair<u8*, PageType> PointerType() const noexcept {
+            const uintptr_t non_atomic_raw = raw;
+            return {ExtractPointer(non_atomic_raw), ExtractType(non_atomic_raw)};
+        }
+
+        [[nodiscard]] uintptr_t Raw() const noexcept {
+            return raw;
```

Comment:
If atomicity is all that matters, this can be:

```suggestion
            return raw.load(std::memory_order_relaxed);
```